sunpy/io/_asdf.py

```diff
@@ -0,0 +1,109 @@
+from pathlib import Path
+from collections import OrderedDict
+
+import numpy as np
+
+import asdf
+
+from sunpy.io._header import FileHeader
+
+__all__ = ["write", "read", "get_header", "get_keys_name"]
+
+def write(fname, data, header, **kwargs):
+    """
+    Take ``(data, header)`` pairs and save it to an ASDF file.
+
+    Parameters
+    ----------
+    fname : `str`
+        File name, with extension.
+    data : `numpy.ndarray`
+        N-dimensional data array.
+    header : `dict`
+        A header dictionary.
+    """
+    map_name = Path(fname)
+    map_name = map_name.name
+    ty = data.dtype
+    data = data.astype(ty)
+    meta = dict(header)
+    with asdf.AsdfFile(copy_arrays=True) as af:
+        af.tree = {map_name : {"meta" : meta, "data" : data}}
+        af.write_to(fname)
+
+
+def read(fname, **kwargs):
+    """
+    Read an ASDF file.
+
+    Parameters
+    ----------
+    filepath : `str`
+        The ASDF file to be read.
+
+    Returns
+    -------
+    `list`
+        A list of ``(data, header)`` tuples.
+    """
+
+    with asdf.open(fname) as af:
+        map_name = get_keys_name(fname)
+        if isinstance(af[map_name],dict):
+            data = af[map_name]["data"].data
+            data_array = np.asarray(data)
+            meta_data= af[map_name]["meta"]
+            meta_data = OrderedDict(meta_data)
+            meta_data = FileHeader(meta_data)
+            return [(data_array,meta_data)]
+        else:
+            data = af[map_name].data
+            meta_data = af[map_name].meta
+            meta_data = FileHeader(meta_data)
+            return [(data,meta_data)]
+
+def get_header(fname):
+    """
+    Read an ASDF file and return just the header(s).
+
+    Parameters
+    ----------
+    fname : `str`
+        The ASDF file to be read.
+
+    Returns
+    -------
+    `list`
+        A list of `sunpy.io._header.FileHeader` headers.
+    """
+    with asdf.open(fname) as af:
+        map_name = get_keys_name(fname)
+        if isinstance(af[map_name], dict):
+            meta_data= af[map_name]["meta"]
+            meta_data = OrderedDict(meta_data)
+            meta_data = FileHeader(meta_data)
+            return [meta_data]
+        else:
+            meta_data = af[map_name].meta
+            meta_data = FileHeader(meta_data)
+            return [meta_data]
+
+
+def get_keys_name(fname):
```

Comment:
I see no mention of this function in the rest of sunpy. Is it required by the io API?

If it is, it's opening `fname` a second time for the uses in this file. Is it's goal to get the first non-"asdf_library"/"history" key in the file? If so this can be done with the already opened file (all the calls above are within a `with asdf.open...` block).

Reply:
i added get_keys_name for handling asdf file which might have different tree name . i tried this in the sense of generalising it

Reply:
I think some consensus for the written expected file structure will need to be reached before considering expanding it to other structures.